File: Sources/JetStreamRpc/Mux.swift
// JetStream RPC — Multiplexer
// Copyright (c) 2024, Sevki <s@sevki.io>
// SPDX-License-Identifier: BSD-3-Clause

// r[impl jetstream.rpc.swift.mux]

import Foundation
import JetStreamWireFormat

/// Client multiplexer for concurrent RPC calls.
public actor Mux<TReq: Framer, TRes: Framer> {
    private let tagPool: TagPool
    private var pending: [UInt16: CheckedContinuation<Frame<TRes>, Error>] = [:]
    private let transport: any Transport<TReq, TRes>

    public init(transport: any Transport<TReq, TRes>, maxConcurrent: UInt16 = 256) {
        self.tagPool = TagPool(maxConcurrent: maxConcurrent)
        self.transport = transport
    }

    /// Start the demux loop (call once).
    public func start() async {
        Task {
            for try await frame in transport.receive() {
                self.dispatch(frame)
            }
        }
    }

    private func dispatch(_ frame: Frame<TRes>) {
        if let continuation = pending.removeValue(forKey: frame.tag) {
            Task { await tagPool.release(frame.tag) }
            continuation.resume(returning: frame)
        }
    }

    /// Send a request and await the response.
    public func rpc(_ msg: TReq) async throws -> Frame<TRes> {
        guard let tag = await tagPool.acquire() else {
            throw FrameError.frameTooSmall(0) // no tags available
        }

        return try await withCheckedThrowingContinuation { continuation in
            pending[tag] = continuation
            Task {
                do {
                    try await transport.send(Frame(tag: tag, msg: msg))
                } catch {
                    if let cont = self.pending.removeValue(forKey: tag) {
                        Task { await self.tagPool.release(tag) }
                        cont.resume(throwing: error)
                    }
                }
            }
        }
    }

    public func close() async throws {
        try await transport.close()
    }
}

File: packages/jetstream_rpc/src/mux.ts
/**
 * r[impl jetstream.rpc.ts.mux]
 */
import type { Frame, Framer } from './frame.js';
import type { Transport } from './transport.js';
import { TagPool } from './tag-pool.js';

export class Mux<TReq extends Framer, TRes extends Framer> {
  private tagPool: TagPool;
  private pending: Map<number, { resolve: (frame: Frame<TRes>) => void; reject: (err: Error) => void }> = new Map();
  private transport: Transport<TReq, TRes>;
  private running = false;

  constructor(transport: Transport<TReq, TRes>, maxConcurrent: number = 256) {
    this.transport = transport;
    this.tagPool = new TagPool(maxConcurrent);
  }

  async start(): Promise<void> {
    this.running = true;
    (async () => {
      for await (const frame of this.transport.receive()) {
        const pending = this.pending.get(frame.tag);
        if (pending) {
          this.pending.delete(frame.tag);
          this.tagPool.release(frame.tag);
          pending.resolve(frame);
        }
      }
    })();
  }

  async rpc(msg: TReq): Promise<Frame<TRes>> {
    const tag = this.tagPool.acquire();
    if (tag === null) throw new Error('no tags available');

    return new Promise((resolve, reject) => {
      this.pending.set(tag, { resolve, reject });
      this.transport.send({ tag, msg }).catch(reject);
    });
  }

  async close(): Promise<void> {
    this.running = false;
    await this.transport.close();
  }
}
